Improved Grid_points and rand_p

ryboselm opened the pull request:
This PR contains only the grid_points and rand_p related changes.
- rand_p function
- grid_points improvement for RealIndexMaps
- grid_points and rand_p tests

ryanlevy commented on test/test_indexmaps.jl:
```diff
@@ -69,4 +69,67 @@ Random.seed!(1234)
+    right = (b - test_gridpoints[end])
+    @test internal >= left && internal >= right
+
+    #second set
+    N = 32
+    a = 0.25
+    b = 0.5
+    test2 = grid_points(s, N, 1; span=[a, b])
+    @test length(test2) == N
+    @test test2[1] >= a
+    @test test2[end] < b
+    internal = (test2[2] - test2[1])
+    left = (test2[1] - a)
+    right = (b - test2[end])
+    @test internal >= left && internal >= right
```
Can each of these commented sets be their own testset?

ryboselm replied:
yes

ryanlevy commented on test/test_indexmaps.jl:
```diff
@@ -69,4 +69,67 @@ Random.seed!(1234)
+    n_grid = 16
+    @test length(grid_points(s, n_grid, 1)) == n_grid
+
+    #test the rand_p() function to see if it succeeds on large L
+    rng = Random.Xoshiro(42)
+    rand_gridpoint1 = rand_p(rng, s)
+    rand_gridpoint2 = rand_p(rng, s, 1)
+    default_rng_gridpoint1 = rand_p(s)
+    default_rng_gridpoint2 = rand_p(s, 1)
```
Nothing is tested here, could you add some `@test`?

ryanlevy reviewed (changes requested):
Thanks Ryan! 

I think the biggest change here is that `rand_p` changes the return value from Float to BigFloat. I have some concerns with this: 

- I'm not sure we have tightened up passing `Number` everywhere (we could maybe just have a test that we can evaluate a basic function on a grid or something)
- grid_points doesn't match this behavior (basically everywhere just does base^L). 
- I'm also not sure we want to enforce BigInt/BigFloat for say `L=4`. We could disbatch on having `base` be BigInt and force the user to deal with this, but I'm not 100% on this at the moment 

For performance reasons I previously swapped `index_value_to_scalar` to return `float(base)^L` instead of `base^L`, but that won't work here due to InexactErrors. We could try to reconcile these choices though

@JoeyT1994  what do you think?

JoeyT1994 commented on src/IndexMaps/realindexmap.jl:
```diff
@@ -75,12 +76,115 @@ function calculate_ind_values(imap::RealIndexMap, xs::Vector, dims::Vector{Int})
+    end
+  end
+
+  #add enforced points in naively
```
Why not just
```
if !isempty(enforced)
  grid_points = sort(vcat(grid_points, enforced_points))` 
end
```
for this logic? I.e. just add them to the end of grid points and re-sort.
I also don't think we need to check the enforced points are in the span: that's something the user can ensure themselves for now

JoeyT1994 commented on src/IndexMaps/realindexmap.jl:
I think we should also add support for multidimensional grid points. 

Like we have a function where we pass a vector of `dimensions` and then zip up the result of `grid_points` called on each of the dimensions

ryboselm replied:
could you elaborate more on what your idea for this would look like?

for example, if you passed in a vector `dimensions = [1,3]`, and suppose that `grid_points(s,1) = [0, 1]` and `grid_points(s,3) = [0.3, 0.4, 0.5]`, would `grid_points(s, [1, 3])` then be something like `[[0, 0.3], [0, 0.4], [0, 0.5], [1, 0.3], [1, 0.4], [1, 0.5]]`?

ryboselm replied:
would it be better to return an array of tuples or an array of arrays here?

I initially leaned towards having an array of arrays because `evaluate` takes in an array to specify the point it should evaluate the function at.

JoeyT1994 commented on src/IndexMaps/realindexmap.jl:
```diff
@@ -75,12 +76,115 @@ function calculate_ind_values(imap::RealIndexMap, xs::Vector, dims::Vector{Int})
+
+  if !exact_grid
+    grid_points = range(span[1], span[2] - (span[2] - span[1]) / N; length=N)
+  else #exact_grid = true
```
I am think this function should follow some more straightforward logic to simplify this code (even if its not quite optimal):

1. Generate an inexact grid using the `range` function
2. If `exact_grid = true` Round all the results to the nearest decimal compatible with the spacing `2^-L` (write a helper function for this) and return  only the unique values you get.

ryboselm replied:
I just made these changes in the simplify grid_points commit

JoeyT1994 reviewed (commented):
Thanks @ryananselm! I added some suggestions, let me know how you get on.

JoeyT1994 commented on src/IndexMaps/realindexmap.jl:
```diff
@@ -75,12 +76,98 @@ function calculate_ind_values(imap::RealIndexMap, xs::Vector, dims::Vector{Int})
+end
+
+function round_to_nearest_exact_point(point::Number, L::Int)
+  if point < 0 || point >= 1
```
I think we can remove this check. I prefer to not have too many error catching exceptions, an error will get thrown eventually anyway

JoeyT1994 commented on test/test_indexmaps.jl:
```diff
@@ -70,3 +70,90 @@ Random.seed!(1234)
+
+  #test the use of default rng
+  default_rng_gridpoint1 = rand_p(s)
+  default_rng_gridpoint2 = rand_p(s, 1)
```
`default_rng_gridpoint2` doesn't seem to be being used

JoeyT1994 commented on test/test_indexmaps.jl:
```diff
@@ -70,3 +70,90 @@ Random.seed!(1234)
+  #test the use of seedeed randomness
+  rng = Random.Xoshiro(42)
+  rand_gridpoint1 = rand_p(rng, s)
+  rand_gridpoint2 = rand_p(rng, s, 1)
```
`rand_gridpoint2` isn't being used in a test

JoeyT1994 commented on test/test_indexmaps.jl:
```diff
@@ -70,3 +70,90 @@ Random.seed!(1234)
+    internal = (test2[2] - test2[1])
+    left = (test2[1] - a)
+    right = (b - test2[end])
+    @test internal >= left && internal >= right
```
What is this testing?

JoeyT1994 commented on src/IndexMaps/realindexmap.jl:
```diff
@@ -75,12 +76,98 @@ function calculate_ind_values(imap::RealIndexMap, xs::Vector, dims::Vector{Int})
+  grid_points = collect(range(span[1], span[2] - (span[2] - span[1]) / N; length=N))
+  if exact_grid
+    for (i, point) in enumerate(grid_points)
+      grid_points[i] = round_to_nearest_exact_point(point, L)
```
One line:
`grid_points = round_to_nearest_exact_point.(grid_points, (L,))`

JoeyT1994 commented on src/IndexMaps/realindexmap.jl:
```diff
@@ -75,12 +76,98 @@ function calculate_ind_values(imap::RealIndexMap, xs::Vector, dims::Vector{Int})
+  end
+  coords = [grid_points(imap, pair[1], pair[2]; kwargs...) for pair in zip(Ns, dims)]
+  gp = Base.Iterators.product(coords...)
+  return [collect(point) for point in gp]
```
`return collect.(gp)`

JoeyT1994 reviewed (commented):
Thanks Ryan! I think we're getting there

ryboselm commented:
For now, to deal with the BigFloat issue, I made it such that it only uses `big()` when L is too large to use a regular Float for. I also cleaned up a lot of the tests and simplified the code according to the suggestions.

ryanlevy commented on src/IndexMaps/realindexmap.jl:
```diff
@@ -75,12 +76,97 @@ function calculate_ind_values(imap::RealIndexMap, xs::Vector, dims::Vector{Int})
+end
+
+function round_to_nearest_exact_point(point::Number, L::Int)
+  return round(point * 2.0^min(63, L)) / 2.0^min(63, L)
```
This assumes base 2, but we shouldnt right? However, for base 2 we can use a trick
```julia
round(ldexp(point,L))/2.0^L
```
That seems to work for any L that can be represented by `point`.

ryboselm replied:
Good point, I generalized this function to take in the base as an argument

ryanlevy commented:
Thanks Ryan, maybe hold off until Joey can comment, but could you specialize anything with `min(63,L)` etc to be a `Float64` (double) instead of `Number`? (Or determine the max L based on the type and base). There's an assumption that the `Number` is Float64 and that `base=2` for 'crossoverL = 63'

ryanlevy commented:
Quick comment, a better way to do the sum that reduces allocations is
```julia
decimal = sum(((i, b),) -> b * ((base * 1.0)^-i), enumerate(bitstring))"
```

ryboselm commented:
thanks! @JoeyT1994 @ryanlevy made a quick fix
